### PR TITLE
Support Kotlin trailing lambda syntax for custom output formatters

### PR DESCRIPTION
**Pull request by vjgarciag96**
Fixes https://github.com/ben-manes/gradle-versions-plugin/issues/584.

**Support Kotlin trailing lambda syntax for custom output formatters**

* This is done by offering a new API in which an output formatter can be specified as org.gradle.api.Action<Result>. Gradle automatically maps Groovy closures and Kotlin lambdas to arguments of that type (https://docs.gradle.org/current/userguide/kotlin_dsl.html#groovy_closures_from_kotlin).

* Internally, there have been some changes done to support this:
  - The type of the property that is used to set an output formatter is `Any`, and this isn't the nicer because tooling can't really hint users of the API with the correct types, and we also have to deal manually with the types of input.
  - The new type we wanted to support (Action<Result>) would run into type erasure problems if we used this existing property and we would need to have code casting the types in an unsafe way (as we can't know if a provided argument is Action<Result> or Action of something else).
  - To sort this out, we have kept the existing property for compatibility reasons (it can still be used to set String, Reporter or Closure output formatters) but we have added a new function to set Action<Result> output formatters. With this we achieve maintaining public API compatibility whilst adding support for the new Action<Result> output formatters with improved type support (internally, safe use of types, and externally, type hinting in tooling and clear public API signature).
  

**Testing**

*Manual*
Manually tested that we can now use the trailing lamdba syntax in Kotlin, and that everything works as before both in Groovy and Kotlin. Also verified that we get type hints in the IDE when using the new API:

image

*Automated*
Added a couple of tests covering the new changes.

### CODE REVIEW

**Comment by ben-manes** [comment]
Thanks!

I believe there was still some remaining problems to work out after the kotlin migration (https://github.com/ben-manes/gradle-versions-plugin/issues/673), so we may not be in a stable state to be able to perform a release. That issue has an answer but I don't think it was followed up on.

Looking at the plugins that use ours, the [gradle-libraries-plugin](https://github.com/fkorotkov/gradle-libraries-plugin/blob/92311acf0cdd7019f7ba7fb55d7c70e98acde9ec/src/main/java/com/fkorotkov/gradle/libraries/tasks/LibrariesBaseTask.java#L44-L50) appears to call some the newly marked `internal` classes (@fkorotkov). It might be kinder to keep them public for the usages to adapt, but otherwise not try to maintain backwards compatibility.

The changes make sense and the code looks good to me. @jaredsburrows can you please review?

**Comment by vjgarciag96** [comment]
@ben-manes Thanks for feedback!

> I believe there was still some remaining problems to work out after the kotlin migration (https://github.com/ben-manes/gradle-versions-plugin/issues/673), so we may not be in a stable state to be able to perform a release. That issue has an answer but I don't think it was followed up on.

Ohhh I see, thanks for the heads-up! I might have a look into those other problems too then.

> Looking at the plugins that use ours, the [gradle-libraries-plugin](https://github.com/fkorotkov/gradle-libraries-plugin/blob/92311acf0cdd7019f7ba7fb55d7c70e98acde9ec/src/main/java/com/fkorotkov/gradle/libraries/tasks/LibrariesBaseTask.java#L44-L50) appears to call some the newly marked internal classes (@fkorotkov). It might be kinder to keep them public for the usages to adapt, but otherwise not try to maintain backwards compatibility.

That makes sense to me, everything is back to public again!

**Comment by jaredsburrows** [comment]
I wonder if converting from `Closure` to `Action` would help resolve the problem we are seeing in #673.

File: gradle-versions-plugin/src/main/kotlin/com/github/benmanes/gradle/versions/updates/DependencyUpdatesTask.kt [hunk 1]
```diff
@@ -12,6 +14,7 @@ import org.gradle.api.tasks.Internal
 import org.gradle.api.tasks.Optional
 import org.gradle.api.tasks.TaskAction
 import org.gradle.util.ConfigureUtil
+import java.lang.IllegalArgumentException
```
Comment (jaredsburrows) [inline gradle-versions-plugin/src/main/kotlin/com/github/benmanes/gradle/versions/updates/DependencyUpdatesTask.kt:17]: This import should not be needed in Kotlin.

File: gradle-versions-plugin/src/main/kotlin/com/github/benmanes/gradle/versions/updates/OutputFormatterArgument.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  /**
+   * An [Action] to provide a custom output formatting. Enables the use of the trailing closure/lambda
+   * syntax for output formatting definition.
+   */
+  class CustomAction(val action: Action<Result>) : OutputFormatterArgument
+
+  companion object {
+    @JvmField
+    val DEFAULT = BuiltIn(formatterNames = "text")
```
Comment (jaredsburrows) [inline gradle-versions-plugin/src/main/kotlin/com/github/benmanes/gradle/versions/updates/OutputFormatterArgument.kt:31]: Does this have to be a `@JvmField`?

**Review by jaredsburrows — APPROVED** [review]

**Comment by vjgarciag96** [comment]
> I wonder if converting from Closure to Action would help resolve the problem we are seeing in https://github.com/ben-manes/gradle-versions-plugin/issues/673.

If the problem there has to do with `rejectVersionIf`, `Action` can't be used because it doesn't have a return type and `rejectVersionIf`'s argument needs to return a boolen. But I think the suggestion from the comment should work (make the `ComponentFilter` interface functional). I can give it a go and put up a PR if it works.

File: gradle-versions-plugin/src/main/kotlin/com/github/benmanes/gradle/versions/updates/DependencyUpdatesTask.kt (hunk 1 above)
Comment (vjgarciag96) [reply gradle-versions-plugin/src/main/kotlin/com/github/benmanes/gradle/versions/updates/DependencyUpdatesTask.kt]: That's right, removed!

File: gradle-versions-plugin/src/main/kotlin/com/github/benmanes/gradle/versions/updates/OutputFormatterArgument.kt (hunk 2 above)
Comment (vjgarciag96) [reply gradle-versions-plugin/src/main/kotlin/com/github/benmanes/gradle/versions/updates/OutputFormatterArgument.kt:30]: I added it to be able to access this from Groovy code in tests. But I have removed it now as it is not necessary in production code.

**Review by ben-manes — APPROVED** [review]
Thanks for contributing! please merge when you’re ready.

**Comment by jaredsburrows** [comment]
Thanks!

**Pull request merged**